src/Lecture2.hs

```diff
@@ -1,170 +1,215 @@
-returning to its lair, the dragon became hungry and ate one of its
-treasure chests by accident.
-
-The guild in the village found a brave knight to slay the dragon!
-As a reward, the knight can take the treasure chest.
-
-Below is the description of the fight and character specifications:
-
-  * A chest contains a non-zero amount of gold and a possible treasure.
-    When defining the type of a treasure chest, you don't know what
-    treasures it stores inside, so your chest data type must be able
-    to contain any possible treasure.
-  * As a reward, the knight takes all the gold, the treasure and experience.
-  * Experience is calculated based on the dragon type. A dragon can be
-    either red, black or green.
-  * Red dragons grant 100 experience points, black dragons — 150, and green — 250.
-  * Stomachs of green dragons contain extreme acid and they melt any
-    treasure except gold. So green dragons have only gold as reward.
-    All other dragons always contain treasure in addition to gold.
-  * Knight tries to slay a dragon with their sword. Each sword strike
-    decreases dragon health by the "sword attack" amount. When the
-    dragon health becomes zero or less, a dragon dies and the knight
-    takes the reward.
-  * After each 10 sword strikes, the dragon breathes fire and decreases
-    knight health by the amount of "dragon fire power". If the
-    knight's health becomes 0 or less, the knight dies.
-  * Additionally, each sword strike decreases "knight's endurance" by one.
-    If a knight's endurance becomes zero, they become tired and are not
-    able to continue the fight so they run away.
-
-Implement data types to describe treasure, knight and dragon.
-And implement a function that takes a knight and a dragon and returns
-one of the three possible fight outcomes.
-
-You're free to define any helper functions.
-
-🕯 HINT: If you find the description overwhelming to implement entirely
-  from scratch, try modelling the problem in stages.
-
-    1. Implement all custom data types without using polymorphism.
-    2. Add @newtype@s for safety where you think is appropriate.
-    3. Encode the fight result as a sum type.
-    4. Add polymorphism.
-    5. Make all invalid states unrepresentable. Think, how you can
-       change your types to prevent green dragons from having any
-       treasure besides gold (if you already haven't done this).
--}
+duplicate [] = []
+duplicate (x : xs) = x : x : duplicate xs
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt 0 (x : xs) = (Just x, xs)
+removeAt pos (x : xs)
+  | pos < 0 = (Nothing, x : xs)
+  | otherwise = second ([x] ++) (removeAt (pos - 1) xs)
```

Comment:
It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible.

```suggestion
  | otherwise = second (x :) (removeAt (pos - 1) xs)
```